File: nodejs/base/api/lib/routing/defaults/routes/default-create-builder.ts
import { v4 as uuidv4 } from "uuid";
import { onResourceCreate, McmaResource, getTableName } from "@mcma/core";
import { DocumentDatabaseTableProvider } from "@mcma/data";

import { getPublicUrl } from "../../../context-variable-provider-ext";
import { DefaultRouteHandlerConfigurator } from "../default-route-handler-configurator";
import { DefaultRouteBuilder } from "../default-route-builder";

export function defaultCreateBuilder<T extends McmaResource>(
    dbTableProvider: DocumentDatabaseTableProvider,
    root: string
): DefaultRouteBuilder<T> {
    return new DefaultRouteBuilder<T>(
        "POST",
        root,
        new DefaultRouteHandlerConfigurator<T>(
            (onStarted?, onCompleted?) =>
                async (requestContext) => {
                    if (onStarted) {
                        const continueRequest = await onStarted(requestContext);
                        if (continueRequest !== undefined && !continueRequest) {
                            return;
                        }
                    }

                    const resource = requestContext.getRequestBody<T>();
                    if (!resource) {
                        requestContext.setResponseBadRequestDueToMissingBody();
                        return;
                    }

                    const resourcePath = root + "/" + uuidv4();

                    onResourceCreate(resource, getPublicUrl(requestContext) + resourcePath);
                    
                    const dbTable = await dbTableProvider.get(getTableName(requestContext));

                    await dbTable.put<T>(resourcePath, resource);
                    
                    if (onCompleted) {
                        await onCompleted(requestContext, resource);
                    }
                    
                    requestContext.setResponseResourceCreated(resource);
                }
        )
    );
}

File: dotnet/base/Mcma.Api/Routing/Defaults/Routes/DefaultCreateBuilder.cs
using System;
using System.Net.Http;
using Mcma.Context;
using Mcma.Data;

namespace Mcma.Api.Routing.Defaults.Routes
{
    public static class DefaultCreateBuilder
    {
        public static DefaultRouteBuilder<TResource> Get<TResource>(IDocumentDatabaseTableProvider dbTableProvider, string root) where TResource : McmaResource
            => 
            new DefaultRouteBuilder<TResource>(
                HttpMethod.Post,
                root,
                new DefaultRouteHandlerConfigurator<TResource>(
                    (onStarted, onCompleted) =>
                        async requestContext =>
                        {
                            // invoke the start handler, if any
                            if (onStarted != null)
                                if (!await onStarted(requestContext))
                                    return;
                            
                            // ensure the body is set
                            var resource = requestContext.GetRequestBody<TResource>();
                            if (resource == null)
                            {
                                requestContext.SetResponseBadRequestDueToMissingBody();
                                return;
                            }

                            var resourcePath = $"{root}/{Guid.NewGuid()}";
                            
                            // initialize the new resource with an ID
                            resource.OnCreate(requestContext.PublicUrlForPath(resourcePath));

                            // put the new object into the table
                            var table = await dbTableProvider.GetAsync(requestContext.TableName());
                            await table.PutAsync(resourcePath, resource);

                            // invoke the completion handler (if any) with the newly-created resource
                            if (onCompleted != null)
                                await onCompleted(requestContext, resource);

                            // return a Created status with the id of the resource
                            requestContext.SetResponseResourceCreated(resource);
                        }));
    }
}